Add Checkout Form components

senadir opened the pull request:
## Introduction
This PR introduces the majority of work in the checkout fields area, it introduces:

### `InputRow`
A component that can hold 1,2 or 3 input fields and lay them out using flex, without the need to pass 50% or size props.

```js
<InputRow>
    <TextInput { ...props } />
    <TextInput { ...props } />
</InputRow>
```

### `TextInput` (to be refactored into base control)
A component that renders a floating label input, in theory, this component is accessible.

```js
const [ email, setEmail ] = useState();
return <TextInput
    id="email"
    label="Email"
    ariaLabel="Shipping Email address"
    screenReaderLabel="Shipping Email address"
    help="We will use this email to update you about your order"
    value={ email }
    onChange={ ( newValue ) => setEmail( newValue ) }
/>
```
![image](https://user-images.githubusercontent.com/6165348/70565510-95c34900-1b92-11ea-86d1-b92134685f2a.png)

![image](https://user-images.githubusercontent.com/6165348/70565488-893ef080-1b92-11ea-97e2-411fff36743c.png)
![image](https://user-images.githubusercontent.com/6165348/70565496-8e9c3b00-1b92-11ea-95e8-016c26a066c9.png)

### `RadioControl`
A component that renders a set of radio buttons with labels, secondary labels, descriptions, and secondary descriptions.

```js
const [ shippingMethod, setShippingMethod ] = useState( 'collect' );
return <RadioControl
    selected={ shippingMethod }
    id="shipping-method"
    onChange={ ( option ) => setShippingMethod( option ) }
    options={ [
        {
            label: 'Click & Collect',
            value: 'collect',
            description: 'Pickup between 12:00 - 16:00 (Mon-Fri)',
            secondaryLabel: 'FREE',
        },
        {
            label: 'Regular shipping',
            value: 'usps-normal',
            description: 'Dispatched via USPS',
            secondaryLabel: '€10.00',
            secondaryDescription: '5 business days',
        },
        {
            label: 'Express shipping',
            value: 'ups-express',
            description: 'Dispatched via USPS',
            secondaryLabel: '€50.00',
            secondaryDescription: '2 business days',
        },
    ] }
/>
```
![image](https://user-images.githubusercontent.com/6165348/70565224-ed14e980-1b91-11ea-8416-989de533d84f.png)

## Inspired by

- `@wordpress/components/radio-control`
- `@wordpress/components/text-control`
- `@automattic/composite-checkout`
- `reakit`

## Problems

- While Gutenberg provides default nice styles for radio, it does not provide them in the frontend, I chose not to tackle the tedious task of customizing checkout and radio fields and focus on other things for now.
- Form management is starting to get out of hand with so many fields and so many states to watch, I suggest we migrate to a hook/slot behavior of defining custom hooks that encapsulates a step `useShippingFieldsStep` that returns the state of that step and fills a slot in the form, another method is to have a global data store and separate those fields to other files and have them all connect to that store.
I'm sure we will figure it.
For reference of how much this can get messy, [composite-checkout demo](https://github.com/Automattic/wp-calypso/blob/master/packages/composite-checkout/demo/index.js).
- Colors are still getting out of hand.

## Questions
- How should we handle extendability for this in the future, this should define how we should define our input fields default values, should we query all fields from the backend and render based on that? or should we have some default fields that exist by default and live in the code? this means we can have slots and fills. js filters or php filters.

## To be tackled in separate PRs
- fixing the checkbox & Radio in the frontend.
- ensuring compatibility with other default themes.
- using Fieldset and Legend in the `formStep` wrapper.

## Future steps
- Introduce a select and phone fields.
- Simplify the code or abstract it more.
- Add sidebar settings to it.

Aljullu commented:
I will try to add my 2-cents to the _Problems_ you listed:

> While Gutenberg provides default nice styles for radio, it does not provide them in the frontend, I chose not to tackle the tedious task of customizing checkout and radio fields and focus on other things for now.

I think that's ok and I would prefer not to load `@wordpress/components` styles in the frontend. In another issue I'm working on the `<Button>` component and Gutenberg styles differ a lot from what we want in the frontend.

> Form management is starting to get out of hand with so many fields and so many states to watch, I suggest we migrate to a hook/slot behavior of defining custom hooks that encapsulates a step `useShippingFieldsStep` that returns the state of that step and fills a slot in the form another method is to have a global data store and separate those fields to other files and have them all connect to that store.

What would be the benefit of this instead of just having a wrapper component that handles the state?

Something like:

```JSX

const Block = () => {
	const [ checkoutFormValues, setCheckoutFormValues ] = useState( {} );
	const setShippingMethod = ( newShippingMethod ) => setCheckoutFormValues( {
		...checkoutFormValues,
		shippingMethod: newShippingMethod.
	} );
	...

	return (
		<CheckoutForm>
			<FormBillingFieldsStep
				shippingMethod={ checkoutFormValues.shippingMethod }
				setShippingMethod={ setShippingMethod }
			/>
			...
		</CheckoutForm>
	);
};
```

> Colors are still getting out of hand.

Right, there is some discussion about it in #1331 and probably we will need to fix it in the entire repo at once, so I wouldn't bother too much about it in this specific PR.

> How should we handle extendability for this in the future, this should define how we should define our input fields default values, should we query all fields from the backend and render based on that? or should we have some default fields that exist by default and live in the code? this means we can have slots and fills. js filters or php filters.

I would leave that for a follow-up as well, it depends on whether we are planning to migrate this block to use inner blocks or not in future versions. If we are using inner blocks, the approach we took for the _All Products_ block might be the best to keep consistency (#1047), if we don't allow inner blocks, we could investigate if SlotFill would be a good option.

Aljullu commented on assets/css/abstracts/_colors.scss:
```diff
@@ -47,5 +47,9 @@ $core-orange: #ca4a1f;
 // @todo: replace those colors with the correct values once we settle on a design system palette.
 $gray-10: #c3c4c7;
 $gray-20: #a7aaad;
+$gray-50: #646970;
```
I just found out this color is not used anywhere, right? Should it be removed or is there any reason to add it here?

senadir replied:
This was removed by mistake in a refactor, it is now restored in 981b072

senadir commented on assets/js/base/components/radio-control/index.js:
```diff
@@ -0,0 +1,119 @@
+						index
+					) => (
+						<label
+							key={ `${ id }-${ index }` }
```
I was hesitant to use that a well but I went with what used in `@wordpress/components/radio-control` I guess we can replace it with the radio value since it should be unique

nerrad replied:
> I was hesitant to use that a well but I went with what used in @wordpress/components/radio-control 

Yikes, that seems like something that should be fixed in `@wordpress/components`.  Value seems a good option here and is the same as what calypso is doing:

https://github.com/Automattic/wp-calypso/blob/master/client/components/forms/form-radios-bar/index.jsx#L20-L40

senadir commented on assets/js/blocks/cart-checkout/checkout/edit.js:
```diff
@@ -14,9 +9,7 @@ const Edit = ( { attributes } ) => {
-			<Disabled>
-				<Block attributes={ attributes } />
-			</Disabled>
+			<Block attributes={ attributes } />
```
I removed them so that we can easily interact with the components within the editor, I think a @todo to remind us to put it back later should be fine?

nerrad replied:
Why do you need to interact with them in the editor?  You could just have a tab that has them available on the frontend for testing interactions?

senadir commented on assets/js/base/components/input-row/index.js:
```diff
@@ -0,0 +1,19 @@
+
+const InputRow = ( { className, children } ) => {
+	return (
+		<div className={ classnames( 'wc-components-input-row', className ) }>
```
This is debatable, there is a lot of overlapping with classes right now, I've gone along and changed to blocks, but I think we should reconsider in the future something predictable and well-aligned if we ever wish to publish those components

senadir commented:
for the sake of this PR not growing too large (it already is) and not getting stale, I think I will wrap any new features here and add them in future iterations, this include sidebar settings and advancing extra fields, I addressed all comments and added empty state behavior, to test normal shipping fields you can set the `shippingMethods` props in `Block.js` to an array with items in it, I expect that we will fill this prop with some HOC or hook later from the data store
